File: 4kyu/sort_binary_tree_by_levels.py
# https://www.codewars.com/kata/52bef5e3588c56132c0003bc

class Node:
    def __init__(self, L, R, n):
        self.left = L
        self.right = R
        self.value = n

def tree_by_levels(tree):
    queue = [tree]
    sorted_tree = []
    while queue:
        node = queue.pop(0)
        if node:
            queue += [node.left, node.right]
            sorted_tree.append(node.value)
    return sorted_tree

if __name__ == '__main__':
    test_cases = [
            (tree_by_levels(Node(Node(None, Node(None, None, 4), 2), Node(Node(None, None, 5), Node(None, None, 6), 3), 1)), [1, 2, 3, 4, 5, 6]), 
            (tree_by_levels(Node(Node(None, Node(None, None, 4), 2), Node(Node(None, None, 5), Node(None, None, 6), 3), 1)), [1, 2, 3, 4, 5, 6]), 
            ]
    for actual, expected in test_cases:
        if actual != expected:
            print(f"expected: {expected}, but was: {actual}")

File: 4kyu/sort_binary_tree_by_levels.hs
-- https://www.codewars.com/kata/52bef5e3588c56132c0003bc--

data TreeNode a = TreeNode {
    left  :: Maybe (TreeNode a),
    right :: Maybe (TreeNode a),
    value :: a
  } deriving (Show)

treeByLevels :: Maybe (TreeNode a) -> [a]
treeByLevels = concat . sort
  where sort Nothing     = []
        sort (Just node) = [value node] : zipWith' (++) [] [] (sort $ left node) (sort $ right node)

zipWith' :: (a -> b -> c) -> a -> b -> [a] -> [b] -> [c]
zipWith' f a b []     []     = []
zipWith' f a b (x:xs) []     = f x b : zipWith' f a b xs []
zipWith' f a b []     (y:ys) = f a y : zipWith' f a b [] ys
zipWith' f a b (x:xs) (y:ys) = f x y : zipWith' f a b xs ys

buildTree :: [a] -> Maybe (TreeNode a)
buildTree l  = fst $ walk $ split 1 l
  where split _ [] = []
        split n l = h : split (2*n) t
          where (h, t) = splitAt n l
        walk [] = (Nothing, [])
        walk ls@([] : _) = (Nothing, ls)
        walk ((h : t) : ls) = (Just $ TreeNode l r h, t : ls'')
          where (l, ls')  = walk ls
                (r, ls'') = walk ls'

main :: IO()
main = do
  print $ treeByLevels $ buildTree $ [1]
  print $ treeByLevels $ buildTree $ [1, 2, 3]
  print $ treeByLevels $ buildTree $ [1, 2, 3, 4, 5, 6, 7]
